rfc/custom-modules-v1.md

```diff
@@ -0,0 +1,456 @@
+---
+title: "Router Custom Module System v1"
+author: Dustin Deus
+date: 2024-08-11
+status: Draft
+---
+
+# Router Custom Module System v1
+
+- **Author:** Dustin Deus
+- **Date:** 2024-08-11
+- **Status:** Draft
+
+## Abstract
+
+This RFC describes an overhaul of the current module system in the router. The new module system is designed to be more flexible and native to GraphQL. It allows developers to hook into the gateway lifecycle as well as outgoing and incoming requests to subgraphs.
+
+## Introduction
+
+As of today, customers can extend the router with custom modules. These modules can be used to change the behavior of the router, add custom logic, or integrate with other systems. The current module system has several limitations that we want to address with this RFC:
+
+- The current module system is not native to GraphQL. It is based on HTTP middleware and does not provide a GraphQL-specific API.
+- The current module system is inconsistent and hard to use. It does not provide a clear API for developers to intercept and modify GraphQL requests and responses.
+- The current module system does not provide a way to create or modify OpenTelemetry data, logs for different parts of the gateway lifecycle.
+- The current module system does not provide a way to interact with the parsed, normalized, and planned GraphQL operation in order to implement custom logic.
+
+Ultimately, custom modules must be self-contained, composable and testable. They should provide a clear API for developers to interact with the gateway and subgraph lifecycle and implement custom logic without having to understand the internal workings of the router or advanced Go programming concepts.
+To briefly explain the decision to use Go as the language for the module system, we have chosen Go because it is a simple and easy-to-learn language that is widely used in the infrastructure and cloud-native ecosystem. You can build custom integration on top production-grade SDK without re-implementing them from scratch. It superiors to scripting languages like Rhai or cross-compiling WebAssembly because it can be easily debugged, profiled, and tested with any modern IDE (VsCode, Goland, etc.). Not part of this RFC, are our ambitions to make the workflow as smooth as possible with a CLI tool that can scaffold, test, and deploy custom modules. In the future, custom modules could be published to a central registry and shared with the community. A brief overview of the workflow is provided at the end of this RFC.
```

Comment:
If we bundle a custom module as "go module" and publish it on GitHub, sharing modules should be straight forward.

Reply:
Yes, this is the idea. See "Module Registration" section.